src/ripple/basics/base_uint.h

```diff
@@ -549,103 +560,75 @@ using uint160 = base_uint<160>;
 using uint256 = base_uint<256>;
 
 template <std::size_t Bits, class Tag>
-inline int
-compare(base_uint<Bits, Tag> const& a, base_uint<Bits, Tag> const& b)
+[[nodiscard]] inline constexpr std::weak_ordering
+operator<=>(base_uint<Bits, Tag> const& lhs, base_uint<Bits, Tag> const& rhs)
 {
-    auto ret = std::mismatch(a.cbegin(), a.cend(), b.cbegin());
+    // It would be nice to use std::lexicographical_compare_three_way()
+    // here but it is not supported on MacOS yet.
 
-    if (ret.first == a.cend())
-        return 0;
+    auto const ret = std::mismatch(lhs.cbegin(), lhs.cend(), rhs.cbegin());
+
+    if (ret.first == lhs.cend())
+    {
+        return std::weak_ordering::equivalent;
+    }
 
     // a > b
     if (*ret.first > *ret.second)
-        return 1;
+    {
+        return std::weak_ordering::greater;
+    }
 
     // a < b
-    return -1;
-}
-
-template <std::size_t Bits, class Tag>
-inline bool
-operator<(base_uint<Bits, Tag> const& a, base_uint<Bits, Tag> const& b)
-{
-    return compare(a, b) < 0;
-}
-
-template <std::size_t Bits, class Tag>
-inline bool
-operator<=(base_uint<Bits, Tag> const& a, base_uint<Bits, Tag> const& b)
-{
-    return compare(a, b) <= 0;
-}
-
-template <std::size_t Bits, class Tag>
-inline bool
-operator>(base_uint<Bits, Tag> const& a, base_uint<Bits, Tag> const& b)
-{
-    return compare(a, b) > 0;
+    return std::weak_ordering::less;
 }
 
-template <std::size_t Bits, class Tag>
-inline bool
-operator>=(base_uint<Bits, Tag> const& a, base_uint<Bits, Tag> const& b)
-{
-    return compare(a, b) >= 0;
-}
-
-template <std::size_t Bits, class Tag>
-inline bool
-operator==(base_uint<Bits, Tag> const& a, base_uint<Bits, Tag> const& b)
-{
-    return compare(a, b) == 0;
-}
-
-template <std::size_t Bits, class Tag>
-inline bool
-operator!=(base_uint<Bits, Tag> const& a, base_uint<Bits, Tag> const& b)
+template <std::size_t Bits, typename Tag>
+[[nodiscard]] inline constexpr bool
+operator==(base_uint<Bits, Tag> const& lhs, base_uint<Bits, Tag> const& rhs)
 {
-    return compare(a, b) != 0;
+    return lhs.data_ == rhs.data_;
 }
```

Comment:
This is all unnecessarily compex, I think. I would reduce it to just two free non-`friend` functions:

```
template <std::size_t Bits, typename Tag>
std::strong_ordering operator<=>(
    base_uint<Bits, Tag> const& lhs,
    base_uint<Bits, Tag>& rhs)
{
    auto const ret = std::mismatch(lhs.cbegin(), lhs.cend(), rhs.cbegin());

    if (ret.first == lhs.cend())
        return std::strong_ordering::equal;

    // a > b
    if (*ret.first > *ret.second)
        return std::strong_ordering::greater;

    // a < b
    return std::strong_ordering::less;
}

template <std::size_t Bits, typename Tag>
bool operator==(
    base_uint<Bits, Tag> const& lhs,
    base_uint<Bits, Tag>& rhs)
{
    return (lhs <=> rhs) == std::strong_ordering::equal;
}
```

At this point, all six relational operators should be correctly available.

Note that I use [`std::strong_ordering`](https://en.cppreference.com/w/cpp/utility/compare/strong_ordering) instead of [`std::weak_ordering`](https://en.cppreference.com/w/cpp/utility/compare/weak_ordering) because the former implies substitutability (i.e. that equivalent values are indistinguishable from each other, which is true for `base_uint<Bits, Tag>`) whereas the latter does not.

Reply:
Thanks. Fixed.

You are right. The return type should be `std::strong_order` in this case.